Add support for grouping tools in plots

mattpap opened the pull request:
Allows to group common tools in plots and figures (in bokehjs). Adds new `Toolbar.group` and `Toolbar.group_types` properties. See the new `example/interaction/tools/tool_grouping.py` example:

[Screencast from 2025-01-31 10-52-08.webm](https://github.com/user-attachments/assets/2b4052c7-5681-4ccd-b440-c62127d227c8)


fixes #5497

mattpap committed:
Add support for grouping tools in plots

mattpap commented:
Tentatively ready for review. I need to add tests.

hoxbro commented on bokehjs/src/lib/core/has_props.ts:
```diff
@@ -260,6 +260,16 @@ export abstract class HasProps extends Signalable() implements Equatable, Printa
     return attrs
   }
 
+  get dirty_attributes(): Attrs {
```
I'm not entirely sure what dirty means. Maybe add a docstring somewhere.

mattpap replied:
Dirtiness is a well defined concept in bokehjs' properties system, though not documented, so I suppose it won't hurt to do it here.

hoxbro commented on bokehjs/src/lib/models/tools/toolbar.ts:
```diff
@@ -122,22 +122,60 @@ export class ToolbarView extends UIElementView {
+                  grouped.set(`${i}`, [tool])
+                }
+              }
+              return [...grouped.values()].map((group) => {
```
Can `Array.from` be used here?

mattpap replied:
You mean `Array.from(iterator, fn)`? Sure. Is it going to be faster? Maybe, though you never know how slow certain native APIs can be.

hoxbro commented on src/bokeh/models/tools.py:
```diff
@@ -396,8 +401,12 @@ def __init__(self, *args, **kwargs) -> None:
+    Whether to group common tools by type.
+    """)
+
+    group_types = List(Enum(ToolName), default=["hover", "tap"], help="""
```
Maybe only add  `hover` for now? Or does the issue ask for taptool too?

mattpap replied:
The issue asked for tool grouping in general and specifically for hover. In any case, I need to resolve issue #3804 first.

hoxbro reviewed (commented):
TapTool does not seem to behave as expected:

https://github.com/user-attachments/assets/ee5972f4-548c-462b-a243-c0f3d8109650


<details>
<summary><code>code</code></summary>

``` python
from bokeh.layouts import column, row
from bokeh.models import CustomJS, Div, HoverTool, Switch, TapTool
from bokeh.plotting import figure, show

plot = figure(width=400, height=400, tools=["pan", "box_select", "save", "reset"])

cr1 = plot.scatter([1, 2, 3], [3, 3, 3], size=30, color="red")
cr2 = plot.scatter([1, 2, 3], [2, 2, 2], size=30, color="green")
cr3 = plot.scatter([1, 2, 3], [1, 1, 1], size=30, color="blue")

plot.add_tools(HoverTool(renderers=[cr1], description="Hover over red"))
plot.add_tools(HoverTool(renderers=[cr2], description="Hover over green"))
plot.add_tools(HoverTool(renderers=[cr3], description="Hover over blue"))

plot.add_tools(TapTool(renderers=[cr1], description="Tap on red"))
plot.add_tools(TapTool(renderers=[cr2], description="Tap on green"))
plot.add_tools(TapTool(renderers=[cr3], description="Tap on blue"))

text = Div(text="Group tools:")
switch = Switch(active=plot.toolbar.group)
switch.js_on_change("active", CustomJS(args=dict(plot=plot), code="""
export default ({plot}, {active: group}) => {
    plot.toolbar.group = group
}
"""))
layout = column([row([text, switch]), plot])

show(layout)
```

</details>

Also, is there a way to group two out of three hover tools? It's a bit hard to tell if this is the case.

mattpap commented:
> TapTool does not seem to behave as expected:

I forgot that you can't have multiple tap tools active at the same time (issue https://github.com/bokeh/bokeh/issues/3804). I suppose I will fix that next.

mattpap commented:
> Also, is there a way to group two out of three hover tools? It's a bit hard to tell if this is the case.

Currently not. Any suggestions regarding the API for this?

hoxbro commented:
> Any suggestions regarding the API for this?

At the top of my head, it would be something like this: 

``` python
plot.add_tools(HoverTool(renderers=[cr1], description="Hover over red", group="one"))
plot.add_tools(HoverTool(renderers=[cr2], description="Hover over green", group="one"))
plot.add_tools(HoverTool(renderers=[cr3], description="Hover over blue", group="two"))
```

The same `group` (it could be another name) for the tool would be merged. If `group` is not added, it would merge all.

mattpap commented:
> The same group (it could be another name) for the tool would be merged. If group is not added, it would merge all.

I was going to do exactly that.

mattpap commented:
I added support for `Tool.group` and removed `"tap"` from `Toolbar.group_types` for now:

[Screencast from 2025-02-04 12-14-49.webm](https://github.com/user-attachments/assets/81784eb2-17e3-448f-a38b-78ac83b40e61)

I also added support for indication of the number of tools being grouped.

hoxbro commented:
> I also added support for indication of the number of tools being grouped.

I think this should be disabled by default (if it is not already). From a user perspective, it would confuse as there is nothing the user can do with the information.

mattpap commented:
> I think this should be disabled by default (if it is not already). From a user perspective, it would confuse as there is nothing the user can do with the information.

This gives the user indication that they work with multiple tools at once. This will be more useful when I will implement implement partial active state (for when only a subset of tool group's tools is active) and a menu with access to individual tools. Note that this doesn't affect proxies in grid plot (etc.) toolbars.

mattpap commented:
I hid the tool tool counter for now. I will revisit this when implementing individual tool access in tool group buttons via a menu.

mattpap committed:
Increase bundle size limit

hoxbro commented on examples/interaction/tools/tool_grouping.py:
```diff
@@ -0,0 +1,38 @@
+cr4 = plot.scatter([1, 2, 3], [2, 2, 2], size=30, color="purple")
+cr5 = plot.scatter([1, 2, 3], [1, 1, 1], size=30, color="yellow")
+
+plot.add_tools(
```
```suggestion
# The description will be shown when hovering over the tool icon
plot.add_tools(
```

hoxbro reviewed (approved):
LGTM

Very happy to see the added comments in the code :+1:
